Fix buoyancy

Thomio-Watanabe opened the pull request:
- Fix the buoyancy behavior when the models are rotated.
- Read the buoyancy as a input parameter, when defined within the gazebo world model.
- Calculates the viscous Drag.

Thomio-Watanabe committed:
Fix buoyancy
test model and link existance
rename local variables
make link a class attribute

Thomio-Watanabe committed:
Add buoyancy as a input parameter
With this modification the user can set a link buoyancy using a world plugin tag.
When the buoyancy is not set, it will be calculated using the model dimensions.

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -28,74 +26,110 @@ namespace gazebo_underwater
-        physics::LinkPtr link = model->GetLink( sdf->Get<std::string>("link_name") ); 
-        applyBuoyancy(link);
-        applyViscousFriction(link);
+        if(link)
```
In which cases could the link not be present ? Shouldn't you make an error and make sure gazebo quits in this case ?

Thomio-Watanabe replied:
This plugin will not find the link if one misspell the link name, i.e. if the link name inside the plugin tag doesn't match the link model name. (I actually need to test two things: if model_name is defined and if model_name name a model loaded in gazebo - todo).
I don't know if I should quit the simulation, because we can add a feature to apply buoyancy to multiple links/models. So, if we are testing the simulation we can comment a link/model inside the world file and still run the simulation. What is your opinion ?

doudou replied:
My opinion is that "silently failing" or just warning in this case is source of trouble. If you comment the link, then comment the buyoancy

Thomio-Watanabe replied:
ok, I'll change the code

Thomio-Watanabe committed:
Test model and link
- test if the model element is defined in world file and if it is loaded in gazebo world
- test if link element is defined in world file and if it is part of the loaded model

Thomio-Watanabe committed:
Calculate the viscous drag

Thomio-Watanabe committed:
Use gazebo Exception to quit simulation.

Thomio-Watanabe commented:
- test model and link elements
- calculate viscous drag
- use gazebo exception

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -28,74 +27,133 @@ namespace gazebo_underwater
-                math::Vector3(0,0,0));
-        viscousDamping = getParameter<math::Vector3>("viscous_damping","dimensionless",
-                math::Vector3(20,30,50));
         waterLevel = getParameter<double>("water_level","meters", 2.0);
```
a water level of 0 would seem like a more intuitive default

Thomio-Watanabe replied:
water level set to 0

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -28,74 +27,133 @@ namespace gazebo_underwater
+        buoyancy = getParameter<double>("buoyancy","N",0);
+        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
+                math::Vector3(0, 0, 0.15));
+        sideAreas = getParameter<math::Vector3>("side_areas","meter2",
```
Can't you use the link's bounding box to compute that ? (I think I mentioned it in the email, but you did not answer)

Thomio-Watanabe replied:
I was speaking with Jan and I think he wants this way.
The same for the volume variable. (because the real robot may have empty spaces inside it)

quams replied:
Am 01.04.2015 um 17:33 schrieb Thomio Watanabe notifications@github.com:

> In src/GazeboUnderwater.cpp:
> 
> > ```
> >      waterLevel = getParameter<double>("water_level","meters", 2.0);
> > ```
> > -        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
> > -        fluidVelocity = getParametermath::Vector3("fluid_velocity","m/s",math::Vector3(0,0,0));
> > -        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
> > -        dragCoefficient = getParametermath::Vector3("drag_coefficient","dimensionless",
> > -                math::Vector3(1,1,1));
> > -        // buoyancy must be the buoyancy when the model is completely submersed
> > -        buoyancy = getParameter<double>("buoyancy","N",0);
> > -        centerOfBuoyancy = getParametermath::Vector3("center_of_buoyancy","meters",
> > -                math::Vector3(0, 0, 0.15));
> > -        sideAreas = getParametermath::Vector3("side_areas","meter2",
> 
> I was speaking with Jan and I think he wants this way.
> The same for the volume variable. (because the real robot may have empty spaces inside it)

Using the bounding box for the calculation only makes sense, if the volume of this box actually represents the volume of water displaced. In most underwater systems that is not entirely the case, especially with a covered open frame like FlatFish (or HROV for that matter).

So I thought its better to force the user to provide displaced Volume and weight as a parameter.

blue skies
    Jan

## 

-- Jan Christian Albiez -- jan@ankerwin.de -- http://www.ankerwin.de/ --
-- segmentation violation in module reality.o ...
                         ... please shutdown your universe and reboot --

doudou replied:
I was referring to the sideAreas parameter, which are geometric properties. I do agree with the volume/weight thing. My comment was really ambiguous ...

Thomio-Watanabe replied:
Jan @quams , this is your call.

doudou replied:
Ping ...

Thomio-Watanabe committed:
Set default water level to 0 m

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -20,82 +20,140 @@ namespace gazebo_underwater
+        if(distanceToSurface > 0 )
+        {
+            math::Vector3 velocityDifference = link->GetWorldCoGLinearVel() - fluidVelocity;
+            math::Vector3 viscousDrag = - 0.5 * densityOfFluid * sideAreas * dragCoefficient *
```
Don't you need to project (ortho) the bounding box w.r.t the direction of the flow ?

Thomio-Watanabe replied:
What do you mean ?

doudou commented:
@Thomio-Watanabe: what would be the difference between the value of sideAreas and what we would get from the bounding box ?

Thomio-Watanabe commented:
When we use bounding box to calculate the side areas we suppose the model is a rectangular box. 
But, likely the real side areas are smaller than the bounding box areas.

doudou commented:
> When we use bounding box to calculate the side areas we suppose the model is a rectangular box. 
> But, likely the real side areas are smaller than the bounding box areas.

Fair enough. However, I really believe that you should at least offer the option of _by default_ using the bounding box. That would make using the plugin that much nicer.

Thomio-Watanabe commented:
Ok, I will work on this

Thomio-Watanabe committed:
Use BoudingBox size to calculate sideAreas by default

Thomio-Watanabe commented:
Use BoudingBox size to calculate sideAreas by default

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -20,82 +20,153 @@ namespace gazebo_underwater
+                gzthrow("GazeboUnderwater: side_areas cannot be (0.0, 0.0, 0.0).");
+        }else{
+            math::Box linkBoudingBox = link->GetBoundingBox();
+            sideAreas = getParameter<math::Vector3>("side_areas","meter2",
```
Isn't that "get parameter and use bounding box by default" ? If yes, why is there the if (hasElement("side_areas")) ?

Thomio-Watanabe replied:
sorry I've gone crazy...

Thomio-Watanabe replied:
In this case should I rewrite history (rebase) or make another commit ?

doudou replied:
I would personally amend (i.e. rewrite the history). You don't need to rebase: since it is the last commit, you can amend it (there's an option for that in git gui)

Thomio-Watanabe commented:
Amending last commit
